### PR TITLE
Support Exports to Multiple ONNX Opset

### PR DESCRIPTION
**Pull request by lara-hdr**
Support exporting multiple ONNX opsets (more specifically opset 10 for now), following the proposal in https://gist.github.com/spandantiwari/99700e60919c43bd167838038d20f353.
And add support for custom ops (merge with https://github.com/pytorch/pytorch/pull/18297).

This PR will be followed by another PR containing the changes related to testing the ops for different opsets.

### CODE REVIEW

**Comment by lara-hdr** [comment]
@houseroad  could you take a look at this? thanks :)

File: test/onnx/test_operators.py [hunk 1]
```diff
@@ -549,6 +549,14 @@ def test_nonzero(self):
         x = torch.tensor([[[2., 2.], [1., 0.]], [[0., 0.], [1., 1.]]], requires_grad=True)
         self.assertONNX(lambda x: torch.nonzero(x), x)
 
+    def test_topk(self):
+        x = torch.arange(1., 6., requires_grad=True)
+        self.assertONNX(lambda x: torch.topk(x, 3), x)
+
+    def test_topk_opset10(self):
```
Comment (houseroad) [inline test/onnx/test_operators.py]: this is a duplicate with test_master_opset, merge two?

**Review by houseroad — COMMENTED** [review]
This is a great effort for supporting multiple onnx opset. Could you fix the failing tests first? I will do a full review later.

File: test/onnx/test_operators.py (hunk 1 above)
Comment (lara-hdr) [reply test/onnx/test_operators.py]: The plan was to refactor the tests in a second PR.
We would split this file to have one for opset9 tests and another for opset10 tests (so we could have multiple tests for opset 10 as well, and we will move both these tests to the opset10 file).
 What is your thought about this?

**Comment by houseroad** [comment]
tests are still failing

**Comment by lara-hdr** [comment]
@houseroad, looking into the failure of test_dist_broadcast_coalesced_gloo, the other tests are passing

**Comment by houseroad** [comment]
rebase to resolve the conflict?

File: test/onnx/test_operators.py (hunk 1 above)
Comment (houseroad) [reply test/onnx/test_operators.py]: I am hesitate to add too many tests rely on expect files. I am thinking to have a new check mechanism, which doesn't rely on expect files.

We can do the following:
1) check the key components in onnx proto, such as ir_version, opset
2) check the target type of node exists in the graph, probably also check the attribute value
3) rely on the onnx checker to do the schema check

Comment (lara-hdr) [reply test/onnx/test_operators.py]: This is possible for operators that have similar schemas for the different opsets. In this case, topk has a different schema between opset 9 and 10. Would we want to have an expect file for each, or just one for one of the opsets?

Comment (spandantiwari) [reply test/onnx/test_operators.py]: I agree that we should reduce the reliance on expect-file based tests. We can discuss how best to do that. For this PR we are incrementally making changes to enable opset 10 dispatch.

**Review by houseroad — CHANGES_REQUESTED** [review]
by default, the exporter should just export the model to the default opset.
If user would like to export to a different opset, it can just specify the target opset using the parameter (we already have it). 

We should not ask the users to call sym_registry.register_version('', 9)

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+# This import monkey-patches graph manipulation methods on Graph, used for the
+# ONNX symbolics
+import torch.onnx.utils
+
+from functools import wraps
+
+
+# EDITING THIS FILE? READ THIS FIRST!
+#
+# - This file is ONLY for ATen operators (e.g., operators that show up in the
```
[the remark this answers is not in the document]
Comment (lara-hdr) [reply torch/onnx/symbolic_helper.py]: I left this block here to avoid duplicating it in all symbolic_opset<version>.py files.
Should I slightly modify it here and reference it in the other files or would it be preferable to copy it in both symbolic_opset9.py and symbolic_opset10.py?

**Comment by lara-hdr** [comment]
The user does not call sym_registry.register_version('', 9).  
sym_registry.register_version is already called in the export functions in utils.py with the specified opset.
So calling torch.onnx.export() is sufficient, and will register the specified function (or default version in case none is specified).

File: test/onnx/test_verify.py [hunk 3]
```diff
@@ -45,6 +47,7 @@ def forward(self, x, y):
 
         x = torch.tensor([1, 2])
         y = torch.tensor([3, 4])
+        torch.onnx.symbolic_registry.register_op('BrokenAdd', BrokenAdd.symbolic, '', 9)
```
Comment (houseroad) [inline test/onnx/test_verify.py]: Why do we need to call register_op here?

File: test/onnx/test_utility_funs.py [hunk 4]
```diff
@@ -38,6 +38,9 @@ def forward(self, x):
                 return b + x
 
         _set_opset_version(9)
+        import torch.onnx.symbolic_registry as sym_registry
+        sym_registry.register_version('', 9)
```
Comment (houseroad) [inline test/onnx/test_utility_funs.py]: what's the purpose of calling register_version('', 9) here?

File: torch/onnx/symbolic_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (houseroad) [reply torch/onnx/symbolic_helper.py]: You can keep most part here, and leave the reference in the symbolic_opset9.py and symbolic_opset10.py

File: test/onnx/test_verify.py (hunk 3 above)
Comment (lara-hdr) [reply test/onnx/test_verify.py]: The symbolic function used here is not registered since it is not in symbolic_opset9/10.py.
So registering the op is required. We could maybe use register_custom_op_symbolic() in utils.py to do it?

File: test/onnx/test_utility_funs.py (hunk 4 above)
Comment (lara-hdr) [reply test/onnx/test_utility_funs.py]: here _set_opset_version() and _model_to_graph() are called instead of _export(), so register_version() is not called.
I will try moving the call to register_version() to avoid this.

File: test/onnx/test_verify.py (hunk 3 above)
Comment (lara-hdr) [reply test/onnx/test_verify.py]: This was actually not needed by modifying onnx.cpp. 
Pushing the change shortly :)

File: test/onnx/test_operators.py [hunk 5]
```diff
@@ -214,6 +214,7 @@ def test_params(self):
         self.assertONNX(lambda x, y: -torch.sigmoid(torch.tanh(x * (x + y))), x, params=(y, ))
 
     def test_symbolic_mismatch(self):
+
```
Comment (houseroad) [inline test/onnx/test_operators.py]: remove this unnecessary change?

File: test/onnx/test_pytorch_onnx_caffe2.py [hunk 6]
```diff
@@ -1584,6 +1592,7 @@ def forward(self, input):
 
     def test_prim_shape(self):
         x = torch.randn(4, 5, requires_grad=True)
+
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_caffe2.py]: remove this unnecessary change?

File: test/onnx/test_verify.py [hunk 7]
```diff
@@ -30,6 +30,7 @@ def assertVerifyExpectFail(self, *args, **kwargs):
         self.assertTrue(False, msg="verify() did not fail when expected to")
 
     def test_result_different(self):
+
```
Comment (houseroad) [inline test/onnx/test_verify.py]: remove this unnecessary change?

File: torch/csrc/jit/passes/onnx.cpp [hunk 8]
```diff
@@ -295,7 +296,7 @@ void BlockToONNX(
     if (func) {
       pyobj = func->get();
     }
-
+
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx.cpp]: remove this unnecessary change?

File: torch/csrc/jit/passes/onnx.cpp [hunk 9]
```diff
@@ -331,6 +332,7 @@ void BlockToONNX(
     // Call the symbolic function
     // Use a little trampoline function so we can give good error messages
     // upon argument mismatch
+    onnx_registry.attr("register_op")(op->name(), pyobj.attr("symbolic"), "", 0);
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx.cpp]: Why pass 0 as the opset_version here? As we discussed in the proposal, we should pass in the target opset, if the target symbolic is missing, we should fall back to the older/previous opset.

File: torch/onnx/__init__.py [hunk 10]
```diff
@@ -6,6 +6,10 @@
 
 ONNX_ARCHIVE_MODEL_PROTO_NAME = "__MODEL_PROTO"
 
+ir_version = 4
```
Comment (houseroad) [inline torch/onnx/__init__.py]: If these variables are only used in the tests, we should add a TODO here, so later, in the exporter, we will also rely on this variables.

File: torch/onnx/symbolic_opset9.py [hunk 11]
```diff
@@ -1,229 +1,24 @@
...
-    global _export_onnx_opset_version
-    if opset_version == _export_onnx_opset_version:
-        return
-    if opset_version in _onnx_stable_opsets + [_onnx_master_opset]:
-        _export_onnx_opset_version = opset_version
-        return
-    raise ValueError("Unsupported ONNX opset version: " + str(opset_version))
 
+# EDITING THIS FILE? READ THIS FIRST!
+# see Note [Edit Symbolic Files]
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: tell user where to find the notes?

File: torch/onnx/symbolic_opset10.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+# This import monkey-patches graph manipulation methods on Graph, used for the
+# ONNX symbolics
+import torch.onnx.utils
+
+from torch.onnx.symbolic_helper import parse_args, _unimplemented
+import torch.onnx.symbolic_opset9
+
+
+# EDITING THIS FILE? READ THIS FIRST!
+# see Note [Edit Symbolic Files]
```
Comment (houseroad) [inline torch/onnx/symbolic_opset10.py]: Tell user where to find the notes?

File: torch/onnx/symbolic_opset10.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import torch
+import torch.onnx
+# This import monkey-patches graph manipulation methods on Graph, used for the
+# ONNX symbolics
+import torch.onnx.utils
+
+from torch.onnx.symbolic_helper import parse_args, _unimplemented
+import torch.onnx.symbolic_opset9
+
```
Comment (houseroad) [inline torch/onnx/symbolic_opset10.py]: Add some comment for the opset detaisl, such as when the opset is supported by which onnx release version, and release date and so on.

File: torch/onnx/symbolic_opset9.py (hunk 11 above)
Comment (houseroad) [reply torch/onnx/symbolic_opset9.py]: Add some comment for the opset detaisl, such as when the opset is supported by which onnx release version, and release date and so on.

File: torch/onnx/utils.py [hunk 14]
```diff
@@ -202,7 +202,8 @@ def _trace(func, args, operator_export_type, return_outs=False):
     # Special case for common case of passing a single Tensor
     if isinstance(args, torch.Tensor):
         args = (args, )
-
+    from torch.onnx.symbolic_helper import _default_onnx_opset_version, _set_opset_version
```
Comment (houseroad) [inline torch/onnx/utils.py]: We should not call `_set_opset_version` on _default_onnx_opset_version.

File: torch/onnx/utils.py [hunk 15]
```diff
@@ -615,13 +619,25 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
                 return new_op_outputs
             else:
                 symbolic_name = 'prim_' + op_name
-                symbolic_fn = getattr(torch.onnx.symbolic, symbolic_name, None)
-                if symbolic_fn is None:
+                is_exportable = sym_registry.is_registered_op(symbolic_name, '', opset_version)
```
Comment (houseroad) [inline torch/onnx/utils.py]: Add a TODO: later we should lift prim's symbolic out.

File: torch/onnx/utils.py [hunk 16]
```diff
@@ -615,13 +619,25 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
...
+                if not is_exportable:
                     warnings.warn("ONNX export failed on primitive operator {}; please report a bug".format(op_name))
-                    return None
+                symbolic_fn = sym_registry.get_registered_op(symbolic_name, '', opset_version)
                 attrs = {k: n[k] for k in n.attributeNames()}
                 return symbolic_fn(g, *inputs, **attrs)
 
+        # custom ops
+        elif sym_registry.is_registered_version(ns, 0):
```
Comment (houseroad) [inline torch/onnx/utils.py]: custom op should as support opset version. i didn't see any reason, we should not support.

**Review by houseroad — CHANGES_REQUESTED** [review]
I didn't see the fallback mechanism when searching op's symbolics.

As we discussed. 

Option 1: Either we put all the symbolic functions into one map symbolics symbolics[domain][opset_version][name] = function

Option 2: We don't create this map explicitly, but actually, will search back if the symbolic doesn't existing in current opset. For example, if you want to export an add operator to opset 10. You cannot find it in opset 10's symbolic, you should find it in opset 9. Please add a such test case.

File: test/onnx/test_onnx_operators.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
+from test_pytorch_common import TestCase, run_tests
```
Comment (houseroad) [inline test/onnx/test_onnx_operators.py]: This file name is confusing... call it test_onnx_opset?

File: test/onnx/test_onnx_opset.py [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+        ops_9 = [{"op_name" : "TopK", "attributes" : [{"name" : "axis", "i" : -1, "type" : 2},
+                 {"name" : "k", "i" : 3, "type" : 2}]}]
+        ops_10 = [{"op_name" : "Constant", "attributes" : [{"name" : "value", "type" : 4}]},
+                  {"op_name" : "Unsqueeze", "attributes" : [{"name" : "axes", "ints" : [0], "type" : 7}]},
+                  {"op_name" : "TopK", "attributes" : [{"name" : "axis", "i" : -1, "type" : 2}]}]
+        ops = {9 : ops_9, 10 : ops_10}
+        x = torch.arange(1., 6., requires_grad=True)
+        test_onnx_operator(MyModule(), x, ops, opset_versions=[9, 10])
+
```
Comment (houseroad) [inline test/onnx/test_onnx_opset.py]: one more line for lint

**Comment by lara-hdr** [comment]
This is doing Option 1 but instead of falling back when searching for the ops, we fall back while registering the ops (in symbolic_registry.py in register_ops_in_version()).

But supporting opset versions for custom ops requires falling back during the search, since the user will register it for a specific version and we do not know which version will be registered later on.
I added functions with the fallback mechanism for custom ops in symbolic_registry.py (get_registered_custom_op()). 

(if we decide to register all the opsets and fallback during the search, get_registered_custom_op() could be the default search function).

File: torch/onnx/symbolic_opset10.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+# see Note [Edit Symbolic Files] in symbolic_helper.py
+
+# This file exports ONNX ops for opset 10
+# Opset 10 is supported by ONNX release 1.5.0
+# release on 04/24/19
+
+
+# Blacklist operators for this opset version.
+# These operators have been updated in ONNX but not re-implemented here.
+# It is very important to blacklist these operators to avoid exporting
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset10.py]: @houseroad, this is the code for the blacklisting.
_black_list_in_opset() is defined in symbolic_helper.py.
These lines will redefine the ops in black_listed_operators  to raise a warning.

I am adding a test test_blacklisted_ops() in test_verify to test this, in https://github.com/pytorch/pytorch/pull/20036.

**Comment by houseroad** [comment]
Lint the code?

File: torch/onnx/symbolic_opset10.py (hunk 19 above)
Comment (houseroad) [reply torch/onnx/symbolic_opset10.py]: https://github.com/onnx/onnx/blob/master/onnx/defs/operator_sets.h#L540

You can find the full list of changed ops in opset 10 here

File: torch/onnx/symbolic_registry.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
+import warnings
+import importlib
+from inspect import getmembers, isfunction
+
+_registry = {}
```
Comment (houseroad) [inline torch/onnx/symbolic_registry.py]: Add some comment about _registry's structure, such as key, value type/content.

File: torch/onnx/symbolic_opset10.py (hunk 19 above)
Comment (lara-hdr) [reply torch/onnx/symbolic_opset10.py]: adding maxpool, averagepool and dropout.
The other ops are not in opset 9.

**Comment by houseroad** [comment]
The failed onnx test is legit: https://circleci.com/api/v1.1/project/github/pytorch/pytorch/1650207/output/104/0?file=true

You may want to add the max_pool in this pr as well.

File: torch/onnx/symbolic_opset10.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    k = g.op("Constant", value_t=torch.tensor(k, dtype=torch.int64))
+    from torch.onnx.symbolic_opset9 import unsqueeze
+    k = unsqueeze(g, k, 0)
+    return g.op("TopK", self, k, axis_i=dim, outputs=2)
+
+
+def _max_pool(name, tuple_fn, ndims, return_indices):
+    @parse_args('v', 'is', 'is', 'is', 'is', 'i')
+    def symbolic_fn(g, input, kernel_size, stride, padding, dilation, ceil_mode):
+        if not stride:
```
Comment (lara-hdr) [inline torch/onnx/symbolic_opset10.py]: I moved the code for maxpool with dilation (and added ceil_mode).
But when I test with onnxruntime, I can see that there is a shape mismatch when dilation != {1} (with the old code as well).

(for example, I get a mismatch with Maxpool1d(kernel_size=3, stride=2, padding=0, dilatoin=2), and input=torch.randn(1,1,4)).


I will investigate the error and submit another PR with the fix if needed.

**Review by houseroad — APPROVED** [review]
Looks good to me. Thanks a lot of this great effort!

**Comment by lara-hdr** [comment]
Thank you @houseroad!
